plasma_framework/contracts/src/exits/payment/PaymentInFlightExitable.sol

```diff
@@ -108,6 +127,70 @@ contract PaymentInFlightExitable is
         emit InFlightExitStarted(msg.sender, startExitData.inFlightTxHash);
     }
 
+    function challengeInFlightExitNotCanonical(ChallengeCanonicityArgs memory args) public payable {
+        // Check if there is an active in-flight exit from this transaction?
+        uint192 exitId = ExitId.getInFlightExitId(args.inFlightTx);
+        PaymentExitDataModel.InFlightExit storage ife = inFlightExits[exitId];
+        require(ife.exitStartTimestamp != 0, "In-fligh exit doesn't exists");
+
+        // Check that the exit is active and in period 1
+        verifyFirstPhaseNotOver(ife);
+
+        // Check if exit's input was spent via MVP exit
+        // If I'm correct we no longer can check any of inputs was previously exited by Std exit
+        // RootChain sets 254-bit in the timestamp to check this
+        /// verifyInputNotSpent(ife);
+
+        // Check that two transactions are not the same
+        require(
+            keccak256(args.inFlightTx) != keccak256(args.competingTx),
+            "The competitor transaction is the same as transaction in-flight"
+        );
+
+        PaymentTransactionModel.Transaction memory inFlightTx = PaymentTransactionModel.decode(args.inFlightTx);
+
+        // Check that shared input owner signes competing transaction
+        require(
+            isSpendingConditionMet(
+                ife.inputs[args.inFlightTxInputIndex].outputGuard,
+                uint256(0), // should not be used
+                inFlightTx.inputs[args.inFlightTxInputIndex],
+                // can we trust the caller or we should extend storage for input outputTypes?
+                args.competingTxInputOutputType,
```

Comment:
@pdobacz @boolafish - do you think it open us to a thread when we allow caller to "remind" us the input type here.
The option is to extend storage when IFE is started and we can verify the types correctness (ehh, or I believe we can)

IMHO this could be a problem when malicious operator will register "always true" predicate here and challenge everything...
Are spending condition contracts quarantined?

Reply:
I think there are potentially two kinds of attack:

1. There is new possible path on spending condition after exit game registered (and quarantined period passed). That is, after exit game registered, operator adds some new spending condition that impacts it. Though potentially we can safe guard by quarantined period, I was thinking how to just make this not possible. I was thinking like freezing the spending condition registry before the ExitGame is registered, and we only need to rely on a single ExitGame quarantined period to safe guard whole exit game. My latest thinking on this is [in here](https://github.com/omisego/plasma-contracts/issues/214#issuecomment-523722519). It seems to me this solution would be simpler and easier and safer.

2. If two spending condition somehow passes with same data input. Then you can pass-in a tx with some output type 1 and try to cheat using output type 2 spending condition predicate. I think we can solve this by each spending condition hardcoding and check the type. And a wrong spending condition would basically just break the exit game, as a result would be safe guarded by quarantined period of exit game.

Reply:
I think I'm unable to really reason about this without the conflated specs :(.

My other generic comment is that we should aim to separate thinking between:
1/ interfaces and code design
2/ contract security

The (2/) should just rely on the `ExitGames` - if a broken one gets introduced, everyone exits. Then those `ExitGames` should be self-contained and include all the necessary code, shrink-wrapped; no registration of SpendingConditions nor other sub-components. We still may have those abstractions, but that's (1/), so their implementations/instances are fixed from PoV of a given `ExitGame`.

The registration could be a feature of maybe some `ExitGame, v13` that will come later and will allow that. Now we're in desperate need for simplicity.

An ExitGame is bound to a tx type, and drives all "exiting actions" that that tx type can perform (currently SE & IFE, as we know them) and sets the rules for them:

The tx type (`tx type 1`) introduces various output types (`output types 1`) that are protected by spending conditions. A new tx type (`tx type 2`) introduced with a new ExitGame2 will introduce new "exiting actions" for `tx type 2`. `tx type 2` will be able to spend `output types 1` according to **their rules**, which should not restrict the spending tx type (we only would do such restriction if that output type was "special", e.g. RC deposit).

But I'm getting ahead of myself, let's discuss this when fleshing out specs. For now it's just a random plea for simplicity

Reply:
> The (2/) should just rely on the ExitGames 

I am full hand support on this! 

For `1/ interfaces and code design`, the main goal here is to avoid re-audit, and thus bringing in the registry though what we need is a predicate factory/provider (which registry by default have such feature) that we can inject/update to new one for new exit game. 

> tx type 2 will be able to spend output types 1 according to their rules, which should not restrict the spending tx type

I am supportive on throwing away spending condition registry if we know how to do update without re-audit the main code. I think such rule on output is fairly easy to be just decided by output type, but how we are going to do with input rules? We need those when IFE.

Reply:
> avoid re-audit

What if we label that as a nice-to-have feature? The more I think about all this, the more I realize allowing for audits of just parts of the contract suite is slippery ground. The same as with the "if ExitGame not nice, then exit" part - the "not nice" is seen holistically anyway, I doubt one could decide only by looking at the new stuff.

> how we are going to do with input rules?

maybe this is the bit that I'm missing to see the problem, I think those rules should be enough to IFE securely. I'm putting this down as point to clarify when doing specs.

Reply:
> What if we label that as a nice-to-have feature? 

cc. @kasima. I totally see what you're seeing lol and I totally agree and is skeptical on "allowing for audits of just parts of the contract suite is slippery ground". But if that works it's quite a lot of saving on money on the other hands.....

> tx type 2 will be able to spend output types 1 according to their rules, which should not restrict the spending tx type

another thing I find it hard to be without considering spending tx type is that how are you going to decode the spending transaction if you don't know the type? I think you mentioned you played around on elixir side, how did you handle that?

Reply:
for future reference - the thread kinda continued here https://github.com/omisego/plasma-contracts/issues/214#issuecomment-524234173, a more suitable spot.

Not sure how to resolve the original question, considering the discussion at hand, yet...

Reply:
I think this PR does not need to address the interface change for Spending Condition of #214.
Back to the original problem, "do you think it open us to a thread when we allow caller to "remind" us the input type here." As for current interface, it does. We basically cannot check the correctness of output type inside spending condition predicate.

Reply:
Created :issue: #230 to address this .